pkg/istio-agent/sds-agent.go

```diff
@@ -208,56 +246,54 @@ type SDSAgent struct {
-	a := &SDSAgent{}
+func NewAgent(proxyConfig *mesh.ProxyConfig, cfg *AgentConfig) *Agent {
+	a := &Agent{
+		proxyConfig: proxyConfig,
+		cfg: cfg,
+	}
+
+	discAddr := proxyConfig.DiscoveryAddress
 
 	a.SDSAddress = "unix:" + LocalSDS
-	a.ClusterID = clusterID
 
-	// If a workload is using file mounted certs, we do not to have to process CA relaated configuration.
+	// If a workload is using file mounted certs, we do not to have to process CA related configuration.
 	if !shouldProvisionCertificates() {
 		log.Info("Workload is using file mounted certificates. Skipping setting CA related configuration")
 		a.FileMountedCerts = true
 		return a
 	}
 
-	a.PilotCertProvider = pilotCertProvider
-	a.OutputKeyCertToDir = outputKeyCertToDir
-
 	_, discPort, err := net.SplitHostPort(discAddr)
 	if err != nil {
 		log.Fatalf("Invalid discovery address %v %v", discAddr, err)
 	}
 
-	if _, err := os.Stat(jwtPath); err == nil && citadel.ProvCert == "" {
-		// If the JWT file exists, and explicit 'prov cert' is not - use the JWT
-		a.JWTPath = jwtPath
-	} else {
-		// If original /etc/certs or a separate 'provisioning certs' (VM) are present, use them instead of tokens
-		certDir := "./etc/certs"
-		if citadel.ProvCert != "" {
-			certDir = citadel.ProvCert
-		}
-		if _, err := os.Stat(certDir + "/key.pem"); err == nil {
-			a.CertsPath = certDir
-		}
-		// If the root-cert is in the old location, use it.
-		if _, err := os.Stat(certDir + "/root-cert.pem"); err == nil {
-			CitadelCACertPath = certDir
-		}
+	// Auth logic for istio-agent to Cert provider:
+	// - if PROV_CERT is set, it'll be included in the TLS context sent to the server
+	//   This is a 'provisioning certificate' - long lived, managed by a tool, exchanged for
+	//   the short lived certs.
+	// - if a JWTPath token exists, will be included in the request.
 
-		if a.CertsPath != "" {
-			log.Warna("Using existing certificate ", a.CertsPath)
-		} else {
-			// Can't use in-process SDS.
-			log.Warna("Missing JWT token, can't use in process SDS ", jwtPath, err)
+	if _, err := os.Stat(a.cfg.JWTPath); err != nil {
+		log.Warna("Missing JWT token ", a.cfg.JWTPath)
+		a.cfg.JWTPath = ""
+	}
+	// If original /etc/certs or a separate 'provisioning certs' (VM) are present,
+	// add them to the tlsContext. If server asks for them and they exist - will be provided.
+	certDir := "./etc/certs"
+	if citadel.ProvCert != "" {
+		certDir = citadel.ProvCert
+	}
+	if _, err := os.Stat(certDir + "/key.pem"); err == nil {
+		a.CertsPath = certDir
+	}
+	if a.CertsPath != "" {
+		log.Warna("Using existing certificate ", a.CertsPath)
+	}
 
-			// TODO do not special case port 15012
-			if discPort == "15012" {
-				log.Fatala("Missing JWT, can't authenticate with control plane. Try using plain text (15010)")
-			}
-			// continue to initialize the agent.
-		}
+	// If the root-cert is in the old location, use it.
+	if _, err := os.Stat(certDir + "/root-cert.pem"); err == nil {
```

Comment:
Right now it's a bit messy - some env variables that I don't fully understand. 

I'm doing a cleanup on a separate PR - this is already too big. 

I think the correct behavior is by default to auto-detect - since different VMs and pods may have different capabilities.
The env - if set - will require a specific behavior.

- if we have the 'trusted' K8S JWT - use it, else
- if we have 'provisioning certs' (i.e. long lived certs) - use them to get the short lived cert, else
- if we have /etc/certs - use them directly (foreign CA), else 
- use the 'untrusted K8S JWT' ( no audience/expiration ) else
- maybe attempt plain text ? 

But let's not add more to this PR...